### PR TITLE
Feature: Image cropping

### LINKED ISSUE #13: Implement the cropping feature
As title mentioned, the cropping feature is missing.

### PR DESCRIPTION
**Pull request by h-alice**
We implemented the image cropping procedure!

```go
img.Then(Crop(<width>, <height>, <algorithm>))
```
Currently the algorithm only supports `center` and `topleft`.
We will add more algorithms in future improvements.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,3 @@
 *.exe
-vendor
\ No newline at end of file
+vendor
+.DS_Store
```
Comment (NatLee) [inline .gitignore:3]: 天
其實還要一個
`desktop.ini`

File: operation/image_cropping.go [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+	ErrCroppingAreaOutOfBound       = errors.New("crop boundary is outside of the original image")
+	ErrAlignmentMethoudNotSupported = errors.New("alignment method not supported")
+)
+
+// Defines a prototype for crop alignment function. (e.g. Center, TopLeft, BottomRight)
+// The first parameter is the boundary of the original image.
+// The second and third parameters are the width and height of the cropping area, respectively.
+type CropAlignment func(image.Rectangle, int, int) image.Rectangle
+
+func CenterAlignment(original_image_boundary image.Rectangle, cropped_width, cropped_height int) image.Rectangle {
```
Comment (NatLee) [inline operation/image_cropping.go:23]: It's good practice to check the crop width and height are positive before proceeding with the cropping operation.
I think negative values could lead to unexpected behavior.

File: operation/image_cropping.go [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+func GetAlignmentMethodByName(name string) (CropAlignment, error) {
+
+	switch strings.ToLower(name) {
+	case "center":
+		return CenterAlignment, nil
+	case "topleft":
+		return TopLeftAlignment, nil
+	default:
+		return nil, ErrAlignmentMethoudNotSupported
```
Comment (NatLee) [inline operation/image_cropping.go]: Is this a typo? :D
```
Methoud -> Method
```

File: operation/test_resources/test_ayaya.jpg
Comment (NatLee) [inline operation/test_resources/test_ayaya.jpg:1]: uwu ayaya

File: operation/test_resources/test_ayaya.png
Comment (NatLee) [inline operation/test_resources/test_ayaya.png:1]: ayaya uwu

File: operation/image_cropping_test.go [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+	for x := 0; x < cropped_img.Bounds().Dx(); x++ {
+		for y := 0; y < cropped_img.Bounds().Dy(); y++ {
+
+			// Calculate the starting point of cropping area.
+			crop_start := image.Point{X: crop_area.Min.X, Y: crop_area.Min.Y}
+
+			// Calculate the current evaluating point on the original image by shifting the starting point of cropping area.
+			current_point_original_image := crop_start.Add(image.Point{X: x, Y: y})
+
+			// NOTE: This is a tricky part. If using At(x, y) to compare the pixel, it may not work as expected.
```
Comment (NatLee) [inline operation/image_cropping_test.go:44]: Why cannot the method work? :'(

File: operation/image_cropping_test.go [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+			}
+		}
+	}
+
+}
+
+func TestInvalidBoundary(t *testing.T) {
+
+	// Load image.
+	img, err := CreateImageFromFile("test_resources/test_ayaya.png")
```
Comment (NatLee) [inline operation/image_cropping_test.go:156]: Can we use `errors.Is` for error comparison to ensure we're correctly identifying the expected error type? :D

- [https://pkg.go.dev/errors](https://pkg.go.dev/errors)

File: operation/image_cropping_test.go [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+		t.Fatalf("Failed to crop image: %v", cropped_img.LastError())
+	}
+
+	// Calculate the cropping area.
+	crop_area := CenterAlignment(img.Image.Bounds(), 11, 15)
+
+	// Check if cropped image bounds are correct.
+	// 14 - int(11 / 2) = 9
+	// 14 - int(15 / 2) = 7
+	if crop_area.Min.X != 9 || crop_area.Min.Y != 7 || crop_area.Max.X != 20 || crop_area.Max.Y != 22 {
```
Comment (NatLee) [inline operation/image_cropping_test.go:125]: Magic `AYAYA` numbers! uwu

**Review by NatLee — APPROVED** [review]
This commit for image cropping functions is well-structured and covers several important scenarios. There are some minor issues that need checking. If they're fine, it's ready to merge. :D

File: .gitignore (hunk 1 above)
Comment (h-alice) [reply .gitignore:3]: 豪

File: operation/test_resources/test_ayaya.jpg
Comment (h-alice) [reply operation/test_resources/test_ayaya.jpg:1]: ayaya

File: operation/test_resources/test_ayaya.png
Comment (h-alice) [reply operation/test_resources/test_ayaya.png:1]: ayaya ayaya!!

File: operation/image_cropping_test.go (hunk 4 above)
Comment (h-alice) [reply operation/image_cropping_test.go:44]: This is strange, it won't return the color as expected.

File: operation/image_cropping.go (hunk 2 above)
Comment (h-alice) [reply operation/image_cropping.go:23]: I think so, I'll add some fix to the issue. : D

File: operation/image_cropping.go (hunk 3 above)
Comment (h-alice) [reply operation/image_cropping.go]: Yes it is. : (

File: operation/image_cropping_test.go (hunk 6 above)
Comment (h-alice) [reply operation/image_cropping_test.go:125]: It's OK since this is just a test case! : D

File: operation/image_cropping_test.go (hunk 5 above)
Comment (h-alice) [reply operation/image_cropping_test.go:156]: I'll do more research of this part.
Commonly, it just use simple compare to determine type of errors.
```go
if err == ErrSomethingNotWork {
    log.Fatelf("This not works.")
}
```

**Comment by h-alice** [comment]
Let's GOOOO.

**Pull request merged**